doc/algorithms.md

```diff
@@ -0,0 +1,727 @@
+    return row
+
+def levenshtein_distance(a: str, b: str):
+    """The standard Levenshtein distance"""
+    # Initialize the row with only insertions
+    row = list(range(len(a) + 1))
+    
+    # Update row for the full string `b`
+    row = update_lev_row(row, a, b)
+
+    # The distance is the last element in the row
+    return row[-1]
+
+```
+
+<!-- Tests 
+```python
+assert levenshtein_distance('kitten', 'sitting') == 3
+assert levenshtein_distance('abc', 'abc') == 0 
+```
+-->
+
+<details>
+<summary>Version with numpy</summary>
+
+```python
+import numpy as np
+
+def update_lev_row_np(row: list | np.ndarray, a: str, b: str):
+    """
+    Takes a `row` from the Levenshtein matrix and computes the following 
+    `len(b)` rows. Returns only the last row.
+    
+    Args:
+        row: A single row from the Levenshtein matrix. It's length must be
+            `len(a) + 1`.
+        a: The hypothesis string
+        b: The reference string. One update will be computed for every word
+            in `b`.
+    """
+    assert len(a) + 1 == len(row), (len(row), len(a), len(b))
+    row2 = np.copy(row)
+    # Iterate over words in b
+    for i, b_word in enumerate(b, start=1):
+        # row2[0] = row[0] + 1  # Deletion
+        np.add(row[0:1], 1, out=row2[0:1])
+    
+        # Iterate over words in a
+        for j, a_word in enumerate(a, start=1):
+            if b_word == a_word:
+                row2[j] = row[j-1]
+            else:
+                np.minimum(
+                    row2[j - 1:j], # Insertion
+                    row[j:j+1], # Deletion
+                    out = row2[j:j+1],
+                )
+                np.minimum(
+                    row2[j:j+1],
+                    row[j - 1:j],   # 
+                    out = row2[j:j+1]
+                )
+                row2[j] += 1
+            
+        row, row2 = row2, row
+    return row
+```
+</details>
+
+## Time-constrained Levenshtein Distance
+
+> [!NOTE]
+> 
+> [View C++ Implementation](../meeteval/wer/matching/levenshtein.h#398) <br>
+> [View C++ Implementation with alignment tracking](../meeteval/wer/matching/levenshtein.h#254)
+
+The time-constrained Levenshtein Distance adds the constraint that two words that are temporally far apart cannot be matched as correct or substituted.
+This can be achieved with a simple overlaps check in the minimum operation.
+
+The search space can be pruned often dramatically by only computing a narrow band through the Levenshtein matrix.
+This pruning is not implemented here to keep this example simple, but the idea is simple: In the loop (a), iterate only over those words in `a` that overlap with `b_word`.
+
+The final implementation is not that simple, though.
+The begin _and_ end times used for this pruning must be monotonically increasing which can be achieved by introducing accumulated minimum and maximum times for the pruning.
+Additionally, the edges of the pruned area have to be treated correctly.
```

Comment:
The final implementation is not that simple, though.
The overhead of the naive implementation is too large to see a good speedup.
The begin _and_ end times used for this pruning must be monotonically increasing, which can be achieved by introducing accumulated minimum and maximum times for the pruning.
Additionally, the edges of the pruned area have to be treated correctly.